File: src/include/misc/StringUtil.php
<?php

namespace misc;

use InvalidArgumentException;

class StringUtil
{
	/**
	 * 
	 * @param string[] $inputArray 
	 * @return string[] 
	 */
	public static function filterTagArray(array $inputArray): array
	{
		// Initialize an empty result array
		$resultArray = array();

		// Loop through each element in the input array
		foreach ($inputArray as $element) {
			// Trim the element and convert it to lowercase
			$filteredElement = strtolower(trim($element));

			// Split the element into multiple elements by space
			$splitElements = preg_split('/\s+/', $filteredElement);

			if ($splitElements === false) {
				$splitElements = [];
			}

			// Loop through the split elements and remove non-alphanumeric characters
			foreach ($splitElements as $splitElement) {
				// Remove non-alphanumeric characters using a regular expression
				$filteredSplitElement = preg_replace('/[^a-z0-9]/', '', $splitElement);

				// Check if the filtered element is not empty and add it to the result array
				if (!empty($filteredSplitElement)) {
					$resultArray[] = $filteredSplitElement;
				}
			}
		}

		return array_unique($resultArray);
	}
	public static function onlySmallLetters(string $input): string
	{
		$output = strtolower($input);
		$output = preg_replace('/[^a-z]/', '', $output) ?? "";
		return $output;
	}
	public static function onlyNumbers(string $input): string
	{
		$output = preg_replace('/[^0-9]/', '', $input) ?? "";
		return $output;
	}
	public static function removeNewline(string $string): string
	{
		return str_replace(array("\n\r", "\n", "\r"), '', $string);
	}

	/**
	 * 
	 * @param string $separator 
	 * @param string $string 
	 * @return string[] 
	 */
	public static function explodeFilterTrim(string $separator, string $string): array
	{
		if (empty($separator)) {
			throw new InvalidArgumentException("Separator string cannot be empty");
		}
		return array_filter(array_map('trim', explode($separator, $string)));
	}


	/**
	 * 
	 * @param string $url 
	 * @param array<string, string> $newParameters 
	 * @return string 
	 */
	public static function addHttpParameters(string $url, array $newParameters): string
	{
		$urlParts = parse_url($url);

		if ($urlParts === false) {
			throw new InvalidArgumentException("Invalid URL provided");
		}

		if (isset($urlParts['query'])) {
			parse_str($urlParts['query'], $params);
		} else {
			$params = array();
		}

		$params = array_merge($params, $newParameters);

		$query = http_build_query($params);

		$scheme = $urlParts['scheme'] ?? 'http';
		$host = $urlParts['host'] ?? '';
		$path = $urlParts['path'] ?? '';

		// Build the final URL
		$output = $query;
		if ($path) {
			$output = $path . '?' . $output;
		}
		if ($host) {
			$output = $scheme . '://' . $host . $output;
		}

		return $output;
	}
}

File: app/misc/string_util.py
import re


class StringUtil:
    @staticmethod
    def filter_tag_array(input_array: list[str]) -> list[str]:
        result: list[str] = []
        for element in input_array:
            filtered = element.strip().lower()
            split_elements = re.split(r"\s+", filtered)
            for split_el in split_elements:
                cleaned = re.sub(r"[^a-z0-9]", "", split_el)
                if cleaned:
                    result.append(cleaned)
        return list(dict.fromkeys(result))  # unique, preserving order

    @staticmethod
    def only_small_letters(input_str: str) -> str:
        return re.sub(r"[^a-z]", "", input_str.lower())

    @staticmethod
    def only_numbers(input_str: str) -> str:
        return re.sub(r"[^0-9]", "", input_str)

    @staticmethod
    def remove_newline(input_str: str) -> str:
        return input_str.replace("\n\r", "").replace("\n", "").replace("\r", "")

    @staticmethod
    def explode_filter_trim(separator: str, string: str) -> list[str]:
        if not separator:
            raise ValueError("Separator string cannot be empty")
        return [p.strip() for p in string.split(separator) if p.strip()]

    @staticmethod
    def add_http_parameters(url: str, new_parameters: dict[str, str]) -> str:
        from urllib.parse import urlparse, urlencode, parse_qs, urlunparse
        parts = urlparse(url)
        existing = parse_qs(parts.query, keep_blank_values=True)
        # Flatten existing params
        flat_existing = {k: v[0] if len(v) == 1 else v for k, v in existing.items()}
        merged = {**flat_existing, **new_parameters}
        new_query = urlencode(merged, doseq=True)
        return urlunparse(parts._replace(query=new_query))
